### DISCUSSION TITLE
ENAMETOOLONG on long slug

### DISCUSSION THREAD

**Discussion by sukhcha-in** [vercel/next.js#32665] [Help]
I'm facing this issue when slug is too long, any way to fix this other than trimming slug?
What is the maximum limit for slug in this case?

Happening in production only, works fine in dev environment.

```
Failed to update prerender files for /en-IN/product/nakpro-perform-whey-protein-concentrate-24g-protein-5-7g-bcaa-4-7g-glutamine-helps-build-your-muscle-size-added-digestive-enzymes-easy-mixing-low-carbs-easy-digesting-fast-absorbing-raw-whey-protein-supplement-powder-for-men-women-athletes-1-kg-unflavoured-30-servings 
[Error: ENAMETOOLONG: name too long, open '/Users/sukh/Documents/Next/phnew/.next/server/pages/en-IN/product/nakpro-perform-whey-protein-concentrate-24g-protein-5-7g-bcaa-4-7g-glutamine-helps-build-your-muscle-size-added-digestive-enzymes-easy-mixing-low-carbs-easy-digesting-fast-absorbing-raw-whey-protein-supplement-powder-for-men-women-athletes-1-kg-unflavoured-30-servings.html'] 
{
  errno: -63,
  code: 'ENAMETOOLONG',
  syscall: 'open',
  path: '/Users/sukh/Documents/Next/phnew/.next/server/pages/en-IN/product/nakpro-perform-whey-protein-concentrate-24g-protein-5-7g-bcaa-4-7g-glutamine-helps-build-your-muscle-size-added-digestive-enzymes-easy-mixing-low-carbs-easy-digesting-fast-absorbing-raw-whey-protein-supplement-powder-for-men-women-athletes-1-kg-unflavoured-30-servings.html'
}
```

**Comment by icyJoseph** [comment]
Isn't this a Linux issue? I think in Linux file names can be up to 255 characters in size. What OS are you using locally?

**Comment by Crinchy** [comment]
Any update on this?

**Reply by icyJoseph** [reply]
Does your OS support more then 255 characters pathnames? I guess one way out would be like a reverse look up map to rewrite incoming urls to shortened versions 🤔

**Reply by Crinchy** [reply]
Im using Ubuntu 22.04.5 LTS. Its not possible to use shortened versions of the url. Even if someone just enter long urls, they can spam my whole application logs.

**Reply by icyJoseph** [reply]
Aha, this is a new take on this. 

So someone requests route with these many characters, and that's causing it to error like this? Do the urls have `.html` extension? can you tell a bit more of the whole setup where this happens?

**Reply by Crinchy** [reply]
Heres a error log of this case:

Failed to update prerender cache for /🫲🙍🏼‍♀️👂🏻👩🏼‍💻💁🏿‍♀️😌🤑📩🍽🧑🏾‍🌾🧼👩🏻‍❤️‍💋‍👩🏻👨🏽‍🦽‍➡️🧓🏾🤹🏿🧑🏻‍❤️‍💋‍🧑🏾🧽🎾🧏🏽‍♂️🪼🪹👩🏿‍⚕️🧑🏿‍💻🪫🫷🏻.png Error: ENAMETOOLONG: name too long, open '/app/.next/server/app/🫲🙍🏼‍♀️👂🏻👩🏼‍💻💁🏿‍♀️😌🤑📩🍽🧑🏾‍🌾🧼👩🏻‍❤️‍💋‍👩🏻👨🏽‍🦽‍➡️🧓🏾🤹🏿🧑🏻‍❤️‍💋‍🧑🏾🧽🎾🧏🏽‍♂️🪼🪹👩🏿‍⚕️🧑🏿‍💻🪫🫷🏻.png.html'
    at async open (node:internal/fs/promises:633:25)
    at async Object.writeFile (node:internal/fs/promises:1207:14)
    at async FileSystemCache.set (/app/node_modules/next/dist/server/lib/incremental-cache/file-system-cache.js:264:13)
    at async IncrementalCache.set (/app/node_modules/next/dist/server/lib/incremental-cache/index.js:400:13)
    at async /app/node_modules/next/dist/server/response-cache/index.js:121:25
    at async /app/node_modules/next/dist/lib/batcher.js:45:32 {
  errno: -36,
  code: 'ENAMETOOLONG',
  syscall: 'open',
  path: '/app/.next/server/app/🫲🙍🏼‍♀️👂🏻👩🏼‍💻💁🏿‍♀️😌🤑📩🍽🧑🏾‍🌾🧼👩🏻‍❤️‍💋‍👩🏻👨🏽‍🦽‍➡️🧓🏾🤹🏿🧑🏻‍❤️‍💋‍🧑🏾🧽🎾🧏🏽‍♂️🪼🪹👩🏿‍⚕️🧑🏿‍💻🪫🫷🏻.png.html'
}

**Reply by Crinchy** [reply]
I have a dynamic route [slug] with ISR enabled and now, when you just hit the website with a long url, the application throws an error because through the linux limit, next.js cant write the file.

**Reply by icyJoseph** [reply]
Does the app crash upon this? What does the end user see? The dreaded 500 Internal Error page?

**Reply by Crinchy** [reply]
The end user is receiving a 404 not found, but in the background, the application is logging this error.

**Comment by diaz-hfc** [comment]
I am seeing this but with a particular scenario, looks like someone is adding some kind of extra text, like this:

`Failed to update prerender cache for /different-is-better-than-better/'Different is Better Than Better' .html']`

I am not sure if this is some kind of attack of bot crawling, but if you look my url is definitely shorter than 256, the extra text is the problem.

**Comment by aapolaakkio** [comment]
This is still a problem. I think Next.js definitely should validate that the path is not too long before trying write the files. While it probably doesn't cause any harm if you limited the amount of logs, it is still quite annoying.

In my case some bot crawlers started trying to crawl paths like: `/fyysikkokilta.fi%25252525252525252525252525252525252525252525252525252525252525252525252525252525252525252525252520%252525252525252525252525252525252525252525252525252525252525252525252525252525252525252525252525281%25252525252525252525252525252525252525252525252525252525252525252525252525252525252525252525252529.gz` which are obviously too long. While I can block the bots by matching the user agent, I would not like to block something like BingBot. And that doesn't block some attacker from maliciously spamming requests with long paths.

As said above, this only happens in production with ISR enabled dynamic routes. Also it seems like the error happens only once for each path.

This is not a Linux specific problem. I think pretty much all operating systems have some sort of limit for the filename or path.

There is actually more info about this in [Turbopack code](https://github.com/vercel/next.js/blob/ca1e1c7d7b2cf6ac040a4f2bda4b5315ab6b439c/turbopack/crates/turbo-tasks-fs/src/lib.rs#L90-L127).
